cannon/mipsevm/multithreaded/mips.go

```diff
@@ -0,0 +1,315 @@
+		m.preemptThread(thread, true)
+		return nil
+	case exec.SysOpen:
+		v0 = exec.SysErrorSignal
+		v1 = exec.MipsEBADF
+	case exec.SysGetAffinity:
+	case exec.SysMadvise:
+	case exec.SysRtSigprocmask:
+	case exec.SysSigaltstack:
+	case exec.SysRtSigaction:
+	case exec.SysPrlimit64:
+	case exec.SysClose:
+	case exec.SysPread64:
+	case exec.SysFstat64:
+	case exec.SysOpenAt:
+	case exec.SysReadlink:
+	case exec.SysReadlinkAt:
+	case exec.SysIoctl:
+	case exec.SysEpollCreate1:
+	case exec.SysPipe2:
+	case exec.SysEpollCtl:
+	case exec.SysEpollPwait:
+	case exec.SysGetRandom:
+	case exec.SysUname:
+	case exec.SysStat64:
+	case exec.SysGetuid:
+	case exec.SysGetgid:
+	case exec.SysLlseek:
+	case exec.SysMinCore:
+	case exec.SysTgkill:
+	case exec.SysSetITimer:
+	case exec.SysTimerCreate:
+	case exec.SysTimerSetTime:
+	case exec.SysTimerDelete:
+	case exec.SysClockGetTime:
+	default:
+		m.Traceback()
+		panic(fmt.Sprintf("unrecognized syscall: %d", syscallNum))
+	}
+
+	exec.HandleSyscallUpdates(&thread.Cpu, &thread.Registers, v0, v1)
+	return nil
+}
+
+func (m *InstrumentedState) mipsStep() error {
+	if m.state.Exited {
+		return nil
+	}
+	m.state.Step += 1
+
+	// If we've completed traversing both stacks for the wakeup check
+	if m.state.Wakeup != exec.FutexEmptyAddr && m.state.TraverseRight && len(m.state.RightThreadStack) == 0 {
+		m.state.TraverseRight = false
+		m.state.Wakeup = exec.FutexEmptyAddr
+		return nil
+	}
+
+	thread := m.state.getCurrentThread()
+	if thread.Exited {
+		m.popThread()
+		m.stackTracker.DropThread(thread.ThreadId)
+		return nil
+	}
+
+	// Search for the first thread blocked by the wakeup call, if wakeup is set
+	// Don't allow regular execution until we resolved if we have woken up any thread.
+	if m.state.Wakeup != exec.FutexEmptyAddr && m.state.Wakeup != thread.FutexAddr {
+		m.preemptThread(thread, !m.state.TraverseRight)
+		return nil
+	}
+
+	// check if thread is blocked on a futex
+	if thread.FutexAddr != exec.FutexEmptyAddr {
+		// if set, then check futex
+		// check timeout first
+		if m.state.Step > thread.FutexTimeoutStep {
+			// timeout! Allow execution
+			m.onWaitComplete(thread, true)
+			return nil
+		} else {
+			m.memoryTracker.TrackMemAccess(thread.FutexAddr)
+			mem := m.state.Memory.GetMemory(thread.FutexAddr)
+			if thread.FutexVal == mem {
+				// still got expected value, continue sleeping, try next thread.
+				m.preemptThread(thread, true)
+				return nil
+			} else {
+				// wake thread up, the value at its address changed!
+				// Userspace can turn thread back to sleep if it was too sporadic.
+				m.onWaitComplete(thread, false)
+				return nil
+			}
+		}
+	}
+
+	if m.state.StepsSinceLastContextSwitch >= exec.SchedQuantum {
+		// Force a context switch as this thread has been active too long
+		msg := fmt.Sprintf("Thread has reached maximum execution steps (%v) - preempting.", exec.SchedQuantum)
+		m.log.Info(msg, "threadId", thread.ThreadId, "threadCount", m.state.threadCount())
```

Comment:
Not sure if this is the right way to add logging here.  Or if I should just be writing to `stdOut`?

Reply:
This will be too verbose and expensive for the current `SchedQuantum` value. We shouldn't log in the VM hot loop. Alternatively, we can wrap this snippet in a `if m.log.Enabled(ctx, log.LevelTrace)` and replace the `log.Info` with a trace log to improve performance.